### PR TITLE
[Fresh] Support DevTools v3

### PR DESCRIPTION
**Pull request by gaearon**
I haven't verified but I think this would fix the issue we've seen with the old version of the global hook.

### CODE REVIEW

File: packages/react-refresh/src/ReactFreshRuntime.js [hunk 1]
```diff
@@ -472,7 +472,8 @@ export function injectIntoGlobalHook(globalObject: any): void {
     // Do the same for any already injected roots.
     // This is useful if ReactDOM has already been initialized.
     // https://github.com/facebook/react/issues/17626
-    hook.renderers.forEach((injected, id) => {
+    const renderers = hook.renderers || hook._renderers; // DevTools v4 || v3
+    renderers.forEach((injected, id) => {
```
Comment (bvaughn) [inline packages/react-refresh/src/ReactFreshRuntime.js]: In v4, [`renderers` is a `Map`](https://github.com/facebook/react/blob/1e1a98942225771f86d8a81b5ac561a9fbff9263/packages/react-devtools-shared/src/hook.js#L275) but in v3 [`_renderers` is an `Object`](https://github.com/facebook/react-devtools/blob/d839081f79f681617df7584a8ffd2f4163fa40b9/backend/installGlobalHook.js#L127). So `.forEach()` wouldn't work on v3. You'd need to do a different loop for v3.

Comment (gaearon) [reply packages/react-refresh/src/ReactFreshRuntime.js]: I admire your diligence.

Comment (bvaughn) [reply packages/react-refresh/src/ReactFreshRuntime.js]: image

**Comment by Mudassar045** [comment]
@gaearon  can possible `hook._renderers === undefined` ?

**Comment by gaearon** [comment]
Yes, `!= null` checks for that.

**Comment by Mudassar045** [comment]
> Yes, `!= null` checks for that.

JavaScript is so fishy... thanks I learned that `undefined == null` returns `true` and `undefined === null` returns `false`

File: packages/react-refresh/src/ReactFreshRuntime.js [hunk 2]
```diff
@@ -472,7 +472,18 @@ export function injectIntoGlobalHook(globalObject: any): void {
     // Do the same for any already injected roots.
     // This is useful if ReactDOM has already been initialized.
     // https://github.com/facebook/react/issues/17626
-    hook.renderers.forEach((injected, id) => {
+    let renderers = hook.renderers;
```
Comment (jamesgeorge007) [inline packages/react-refresh/src/ReactFreshRuntime.js]:
```suggestion
    let { renderers}  = hook;
```

Comment (jamesgeorge007) [reply packages/react-refresh/src/ReactFreshRuntime.js]: I'm a big fan of destructuring assignment approach. The former code snippet appears kinda redundant atleast IMO.

Comment (jamesgeorge007) [inline packages/react-refresh/src/ReactFreshRuntime.js]:
```suggestion
    let { renderers } = hook;
```

Comment (jamesgeorge007) [reply packages/react-refresh/src/ReactFreshRuntime.js]: I'm a big fan of destructuring assignment approach. The former code snippet appears kinda redundant atleast IMO.

**Comment by gaearon** [comment]
Flow want something. I'll get back to this maybe.

**Pull request closed without merging**